lib/workload/stateless/filemanager/filemanager/src/database/mod.rs

```diff
@@ -2,31 +2,83 @@
 //!
 
 use async_trait::async_trait;
+use sqlx::postgres::PgConnectOptions;
 use sqlx::PgPool;
+use std::borrow::Cow;
+use tracing::debug;
 
 use crate::env::read_env;
 use crate::error::Result;
 use crate::events::EventSourceType;
 
 pub mod aws;
 
+/// A trait which can generate database credentials.
+#[async_trait]
+pub trait CredentialGenerator {
+    /// Generate the password used to connect to the database.
+    async fn generate_password(&self) -> Result<String>;
+}
+
 /// A database client handles database interaction.
 #[derive(Debug)]
-pub struct Client {
-    pool: PgPool,
+pub struct Client<'a> {
+    // Use a Cow here to allow an owned pool or a shared reference to a pool.
+    pool: Cow<'a, PgPool>,
```

Comment:
> Also higher amount of explicit lifetimes <'a> in this changeset, what motivated the compiler to have them defined?

@brainstorm the compiler requires this because I'm using a `Cow` :cow: here which is sort of like a shared-reference smart pointer that can store an owned or reference type for the database pool. This means that everything that uses the database `Client` now has to have a lifetime defined. 

It would be cool if lifetimes could be elided in structs too but I don't think this is planned: https://github.com/rust-lang/rfcs/blob/master/text/0141-lifetime-elision.md#lifetime-elision-in-structs